### DISCUSSION TITLE
如何在form-data请求中正确解析array类型参数？

### DISCUSSION THREAD

**Discussion by kaesi0** [midwayjs/midway#3967] [Q&A] [answered]
### Describe the problem（描述问题）

在/addTag路由中需要设置数组类型的别名参数alias以及上传文件，所以需要以form-data格式发起请求

我的路由是
` @Post('/addTag')
  async addTag(@Files() files, @Fields() fields: AddTagDTO) {
    return await this.tagService.addTag(fields, files);
  }`

AddTagDTO：
`export class AddTagDTO {
  @Rule(RuleType.string().min(2).max(10).required())
  name: string;

  @Rule(RuleType.array().items(RuleType.string()).single())
  alias: string[];
`
当我以form-data形式发起请求时，比如
![image](https://github.com/user-attachments/assets/dd67b535-6c78-40a9-bc21-8fc64065e60a)
在服务端解析后的参数中，alias始终只有最后一个值，也就是“别名1”被覆盖了，这可能与服务端的参数解析设定有关，但不清楚是Fields装饰器的bug还是koa-bodyparser的配置问题，如果是koa-bodyparser的配置问题，在哪里能修改？

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(3.16.5)
✓ @midwayjs/bootstrap(3.16.2)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/cache-manager(3.16.5)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.16.2)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(3.16.5)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.16.5)
✓ @midwayjs/jwt(3.16.5)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.16.5)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(3.16.5)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(3.16.5)
✓ @midwayjs/validate(3.16.5)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.16.5)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by czy88840616** [comment, marked as answer]
当前的确在同名 field 时会直接覆盖，可能得修复处理一下。

**Reply by kaesi0** [reply]
应该是默认的body-parser中间件并不支持解析multipart bodies：https://www.npmjs.com/package/body-parser 我关闭了默认的bodyParser，使用koa-body中单件替换，但是在开启了upload组件的情况下，如果设置koa-body的multipart==true，请求会卡死，应该是与upload组件冲突了，关闭upload组件上传文件和解析数组参数都正常了，但又失去了upload组件好用的自动清理，限制文件格式等功能，请修复一下，如果有修复发布请通知。感谢

**Reply by czy88840616** [reply]
bodyParser和 multipart 是两个不同的组件和流程，bodyParser 不会处理上传逻辑。

**Reply by czy88840616** [reply]
应该是这个地方https://github.com/midwayjs/midway/blob/main/packages/upload/src/parse.ts#L121

**Reply by kaesi0** [reply]
恩我理解bodyParser和 multipart 是两个不同的组件和流程。我更喜欢使用官方的默认的组件而不是去替换，官方的upload组件很好用，但是现在确实在form-data请求中即要上传文件，又需要传递数组参数，虽然可以以字符串（比如逗号分割）的方式传递数组参数，通过 自定义的参数验证规则或者管道整理成数组，但总归不如通过组件和配置就完成功能。

**Reply by kaesi0** [reply]
> 应该是这个地方https://github.com/midwayjs/midway/blob/main/packages/upload/src/parse.ts#L121

是说这里的问题导致koa-body和upload冲突么？

**Reply by czy88840616** [reply]
就是这里没处理同名 field

**Reply by kaesi0** [reply]
大佬厉害，这么快定位问题，只有靠你们修复了，这里确实搞不懂了
